### PR TITLE
Add new matric "spatial distortion index"

### PR DESCRIPTION
**Pull request by ywchan2005**
## What does this PR do?

Add new metric 'D_s' (spatial distortion index)
(being part of #799 )

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--2260.org.readthedocs.build/en/2260/

### CODE REVIEW

File: src/torchmetrics/functional/image/d_s.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Tuple
+
+import torch
+from kornia.filters import filter2d
```
Comment (Borda) [inline src/torchmetrics/functional/image/d_s.py:18]: You shall use conditional or lazy import... see for example: #2215

Comment (ywchan2005) [reply src/torchmetrics/functional/image/d_s.py]: seems to work now. thanks

File: docs/source/image/spatial_distortion_index.rst [hunk 2]
```diff
@@ -0,0 +1,21 @@
+.. customcarditem::
+   :header: Spatial Distortion Index
+   :image: https://pl-flash-data.s3.amazonaws.com/assets/thumbnails/image_classification.svg
+   :tags: Image
+
+.. include:: ../links.rst
+
+#########################
+Spatial Distortion Index
+#########################
```
Comment (SkafteNicki) [inline docs/source/image/spatial_distortion_index.rst:10]:
```suggestion
########################
Spatial Distortion Index
########################
```

File: src/torchmetrics/functional/image/d_s.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    if pan_lr is not None:
+        pan_lr_h, pan_lr_w = pan_lr.shape[-2:]
+        if pan_lr_h != ms_h:
+            raise ValueError(
+                f"Expected `ms` and `pan_lr` to have the same height. Got ms: {ms_h} and pan_lr: {pan_lr_h}."
+            )
+        if pan_lr_w != ms_w:
+            raise ValueError(
+                f"Expected `ms` and `pan_lr` to have the same width. Got ms: {ms_w} and pan_lr: {pan_lr_w}."
+            )
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/image/d_s.py:159]: all this is just input checking so it should probably just be moved to the `_update` method

File: src/torchmetrics/functional/image/d_s.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+    if pan_lr is None:
+        from torchvision.transforms.functional import resize
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/image/d_s.py:162]:
```suggestion
        if not _TORCHVISION_AVAILABLE:
            raise ValueError("When `pan_lr` is not provided as input to metric Spatial distortion index, torchvision should be installed. Please install with `pip install torchvision` or `pip install torchmetrics[image]`.")
        from torchvision.transforms.functional import resize
```

File: src/torchmetrics/image/d_s.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        """Compute and returns spatial distortion index."""
+        preds = dim_zero_cat(self.preds)
+        ms = dim_zero_cat(self.ms)
+        pan = dim_zero_cat(self.pan)
+        pan_lr = dim_zero_cat(self.pan_lr) if len(self.pan_lr) > 0 else None
+        target = {
+            "ms": ms,
+            "pan": pan,
+            **({"pan_lr": pan_lr} if pan_lr is not None else {}),
+        }
```
Comment (SkafteNicki) [inline src/torchmetrics/image/d_s.py:136]:
```suggestion
        target = {"ms": ms, "pan": pan}
        target.update({"pan_lr": pan_lr} if pan_lr is not None else {})
```

File: src/torchmetrics/functional/image/d_s.py (hunk 3 above)
Comment (ywchan2005) [reply src/torchmetrics/functional/image/d_s.py]: thanks for the suggestion. i've moved the checking of tensor to `_update` method

File: src/torchmetrics/functional/image/d_s.py [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+def _spatial_distortion_index_update(preds: Tensor, target: Dict[str, Tensor]) -> Tuple[Tensor, Dict[str, Tensor]]:
+    """Update and returns variables required to compute Spatial Distortion Index.
+
+    Args:
+        preds: High resolution multispectral image.
+        target: A dictionary containing the following keys:
+
+            - ``'ms'``: low resolution multispectral image.
+            - ``'pan'``: high resolution panchromatic image.
+            - ``'pan_lr'``: (optional) low resolution panchromatic image.
```
Comment (Borda) [inline src/torchmetrics/functional/image/d_s.py:35]: why not have rather ms, pan, pan_lr as arguments?

File: src/torchmetrics/functional/image/d_s.py [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    """
+    if len(preds.shape) != 4:
+        raise ValueError(f"Expected `preds` to have BxCxHxW shape. Got preds: {preds.shape}.")
+    if "ms" not in target or "pan" not in target:
+        raise ValueError(f"Expected `target` to have keys ('ms', 'pan'). Got target: {target.keys()}")
+    for name, t in target.items():
+        if preds.dtype != t.dtype:
+            raise TypeError(
+                f"Expected `preds` and `{name}` to have the same data type. "
+                "Got preds: {preds.dtype} and {name}: {t.dtype}."
```
Comment (Borda) [inline src/torchmetrics/functional/image/d_s.py:65]:
```suggestion
                f"Expected `preds` and `{name}` to have the same data type."
                " Got preds: {preds.dtype} and {name}: {t.dtype}."
```

File: src/torchmetrics/functional/image/d_s.py [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+            )
+    for name, t in target.items():
+        if len(t.shape) != 4:
+            raise ValueError(f"Expected `{name}` to have BxCxHxW shape. Got {name}: {t.shape}.")
+    for name, t in target.items():
+        if preds.shape[:2] != t.shape[:2]:
+            raise ValueError(
+                f"Expected `preds` and `{name}` to have same batch and channel sizes. "
+                "Got preds: {preds.shape} and {name}: {t.shape}."
```
Comment (Borda) [inline src/torchmetrics/functional/image/d_s.py:74]:
```suggestion
                f"Expected `preds` and `{name}` to have the same batch and channel sizes."
                " Got preds: {preds.shape} and {name}: {t.shape}."
```

File: src/torchmetrics/functional/image/d_s.py [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+                f"Expected `ms` and `pan_lr` to have the same width. Got ms: {ms_w} and pan_lr: {pan_lr_w}."
+            )
+
+    return preds, target
+
+
+def _spatial_distortion_index_compute(
+    preds: Tensor,
+    target: Dict[str, Tensor],
+    p: int = 1,
```
Comment (Borda) [inline src/torchmetrics/functional/image/d_s.py:120]: lets have name of at least 3 letters

File: src/torchmetrics/functional/image/d_s.py [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+        m2[i] = universal_image_quality_index(preds[:, i : i + 1], pan[:, i : i + 1])
+    diff = (m1 - m2).abs() ** p
+    return reduce(diff, reduction) ** (1 / p)
+
+
+def spatial_distortion_index(
+    preds: Tensor,
+    target: Dict[str, Tensor],
+    p: int = 1,
+    ws: int = 7,
```
Comment (Borda) [inline src/torchmetrics/functional/image/d_s.py:200]:
```suggestion
    window_size: int = 7,
```

File: src/torchmetrics/functional/image/d_s.py (hunk 9 above)
Comment (Borda) [reply src/torchmetrics/functional/image/d_s.py:120]:
```suggestion
    norm_order: int = 1,
```

**Review by Borda — APPROVED** [review]
overall looks good, just some API sugegstions

File: src/torchmetrics/functional/image/d_s.py (hunk 6 above)
Comment (ywchan2005) [reply src/torchmetrics/functional/image/d_s.py:35]: The interface is common to other metrics, including the functional and class interface. Changing the interface affects less on the `src` side, but more on the `tests` side. It requires the helper functions (e.g. `MetricTester.run_functional_metric_test`) to support such variety of arguments

Comment (Borda) [reply src/torchmetrics/functional/image/d_s.py:35]: I think that for private function we can be more specific if it improves readability

**Review by SkafteNicki — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
